### PR TITLE
add function "last" for supporting last sample in a time window

### PR DESCRIPTION
**Pull request by sherali42**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?

add `last` function for supporting last sample in a time window. This implementation ensures that timestamp of last sample is after start-time and before end-time of the window.

### CODE REVIEW

**Comment by sherali42** [comment]
> Hey @sherali42 .... you don't need to pass in an extra parameter Window. There are already too many parameters to the `addChunks`. To check the number of samples in the window, you have to query the Timestamp vector reader and check the start and end row numbers. The window is not necessary.
> Also, the ChunkedWindowIterator does not use Windows at all, so I'm not sure how that is working.
> 
> I think this PR can be much, much much smaller. Just use the existing LastSampleRangeFunction which already does the right thing... we don't need to add another custom RangeFunction. It also already takes into account all the mechanics behind NaN and if there are no samples.

@velvia I have updated the code, could you please take a look now.

File: query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala [hunk 1]
```diff
@@ -77,6 +77,8 @@ trait RangeFunction extends BaseRangeFunction {
             window: Window,
             sampleToEmit: TransientRow,
             queryConfig: QueryConfig): Unit
+
```
Comment (velvia) [inline query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala]: Remove extra blank lines?

File: query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala [hunk 2]
```diff
@@ -394,14 +398,35 @@ object LastSampleFunction extends RangeFunction {
...
+  override def apply(startTimestamp: Long,
+            endTimestamp: Long,
+            window: Window,
+            sampleToEmit: TransientRow,
+            queryConfig: QueryConfig): Unit = {
+    if (window.size > 1)
+      throw new IllegalStateException(s"Window had more than 1 sample. Possible out of order samples. Window: $window")
+    // if no sample found in the window, set the value to NaN
+    if (window.size == 0 || startTimestamp > window.head.getLong(0)) {
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala]: `startTimestamp > window.head.getLong(0)` is not needed. Always true.

File: query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala [hunk 3]
```diff
@@ -496,6 +521,30 @@ class LastSampleChunkedFunctionD extends LastSampleChunkedFuncDblVal() {
   }
 }
 
+// Return the last sample within the window
+class LastSampleInWindowChunkedFunctionD extends LastSampleChunkedFunctionD {
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala]: The abstract class and implementation for double and histogram would be useful. Lets quickly add them and move on.

Comment (sherali42) [reply query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala]: i have added the stack to support double, long and histo types

File: query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala (hunk 2 above)
Comment (sherali42) [reply query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala]: looks like this condition is needed, as we add last sample before start of the window. This is where the logic is to include the last sample outside the window `SlidingWindowIterator.shouldAddCurToWindow`

**Pull request merged**